File: src/main/java/net/azisaba/yukitexture/JedisBox.java
package net.azisaba.yukitexture;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPoolConfig;

import java.io.Closeable;
import java.util.Objects;

public final class JedisBox implements Closeable {
    private final JedisPool jedisPool;

    public JedisBox(@NotNull String hostname, int port, @Nullable String username, @Nullable String password) {
        this.jedisPool = createPool(hostname, port, username, password);
    }

    @Contract(pure = true)
    @NotNull
    public JedisPool getJedisPool() {
        return jedisPool;
    }

    @Override
    public void close() {
        getJedisPool().close();
    }

    @Contract("_, _, _, _ -> new")
    public static @NotNull JedisPool createPool(@NotNull String hostname, int port, @Nullable String username, @Nullable String password) {
        Objects.requireNonNull(hostname, "hostname");
        if (username != null && password != null) {
            return new JedisPool(hostname, port, username, password);
        } else if (password != null) {
            return new JedisPool(new JedisPoolConfig(), hostname, port, 3000, password);
        } else if (username != null) {
            throw new IllegalArgumentException("password must not be null when username is provided");
        } else {
            return new JedisPool(new JedisPoolConfig(), hostname, port);
        }
    }
}

File: src/main/kotlin/net/azisaba/yukitexture/JedisBox.kt
package net.azisaba.yukitexture

import java.io.Closeable
import java.util.Objects
import org.jetbrains.annotations.Contract
import redis.clients.jedis.JedisPool
import redis.clients.jedis.JedisPoolConfig

class JedisBox(hostname: String, port: Int, username: String?, password: String?) : Closeable {
    internal val jedisPool: JedisPool = createPool(hostname, port, username, password)

    @Contract(pure = true)
    fun getJedisPool(): JedisPool {
        return jedisPool
    }

    override fun close() {
        getJedisPool().close()
    }

    companion object {
        @Contract("_, _, _, _ -> new")
        fun createPool(hostname: String, port: Int, username: String?, password: String?): JedisPool {
            Objects.requireNonNull<String?>(hostname, "hostname")
            if (username != null && password != null) {
                return JedisPool(hostname, port, username, password)
            } else if (password != null) {
                return JedisPool(JedisPoolConfig(), hostname, port, 3000, password)
            } else require(username == null) { "password must not be null when username is provided" }
            return JedisPool(JedisPoolConfig(), hostname, port)
        }
    }
}